ssl/statem/statem_srvr.c

```diff
@@ -3138,7 +3138,7 @@ static int tls_process_cke_ecdhe(SSL_CONNECTION *s, PACKET *pkt)
         }
 
         if (EVP_PKEY_set1_encoded_public_key(ckey, data, i) <= 0) {
-            SSLfatal(s, SSL_AD_INTERNAL_ERROR, ERR_R_EC_LIB);
+            SSLfatal(s, SSL_AD_ILLEGAL_PARAMETER, ERR_R_EC_LIB);
```

Comment:
This PR brings to light a more interesting potential issue than that which alert is raised.

* In the DHE case the function `EVP_PKEY_set1_encoded_public_key` is expected to return a boolean value (0 or non-zero)
* In the ECDHE case the **same** function is expected to return a positive value for success and a zero or negative value for failure.

One convention used in some places in the code is that 0 is expected failure, while negative is unexpected failure, and they can be treated differently, allowing, say, malloc failures to still report internal error, while actual problems with the input might report the illegal parameter alert.

The manpage says:
```
RETURN VALUES
    EVP_PKEY_set1_encoded_public_key() returns 1 for success and 0 or a
    negative value for failure.
```
So the DHE case is not strictly correct, but the implementation always returns 0 or 1 for legacy keys: https://github.com/openssl/openssl/blob/0958f5a5bc47e2cca907a3bfaf14059c020324fd/crypto/evp/p_lib.c#L1386-L1404
Leaving just provided keys as a potential source of negative return values, but that ultimately takes us to: https://github.com/openssl/openssl/blob/0958f5a5bc47e2cca907a3bfaf14059c020324fd/crypto/evp/keymgmt_meth.c#L485 so up to whatever the provider does. This time documentation states: https://github.com/openssl/openssl/blob/0958f5a5bc47e2cca907a3bfaf14059c020324fd/doc/man7/provider-keymgmt.pod?plain=1#L478-L479

So, bottom line, the negative values are historical, and at this point probably can't be brought back without breaking the API.
So, in the end, no way to distinguish input errors from resource errors. Oh, well...

@t8m should we just fix the nits with the 3rd argument and merge without waiting for the author?

Reply:
> @t8m should we just fix the nits with the 3rd argument and merge without waiting for the author?

If you want to, you can open a new PR. Or push a fixup commit to this author's branch.

Reply:
Done.